apps/pipelines/nodes/base.py

```diff
@@ -0,0 +1,93 @@
+from abc import ABC
+from collections.abc import Callable, Sequence
+from typing import Annotated, Any, TypedDict
+
+from langchain.prompts import BasePromptTemplate
+from langchain_core.messages import BaseMessage
+from langchain_core.prompt_values import StringPromptValue
+from langchain_core.runnables import (
+    Runnable,
+)
+from pydantic import BaseModel
+from pydantic_core import ValidationError
+
+from apps.pipelines.exceptions import PipelineNodeBuildError
+from apps.pipelines.graph import Node
+from apps.pipelines.logging import PipelineLoggingCallbackHandler
+
+
+def add_messages(left: list, right: list):
+    # Could probably log here
+    return left + right
+
+
+class PipelineState(TypedDict):
+    messages: Annotated[Sequence[Any], add_messages]
+
+
+class PipelineNode(BaseModel, ABC):
+    """Pipeline node that is either a single or a combination of Langchain Runnables
+
+    Define required parameters as typed fields. Compose the runnable in `get_runnable`
+
+    Example:
+        class FunNode(PipelineNode):
+            required_parameter_1: CustomType
+            optional_parameter_1: Optional[CustomType] = None
+
+            def get_runnable(self, input) -> str:
+                if self.optional_parameter_1:
+                    return PromptTemplate.from_template(template=self.required_parameter_1)
+                else:
+                    return LLMResponse.build(node)
+
+       class FunLambdaNode(PipelineNode):
+            required_parameter_1: str
+
+            def get_runnable(self, node: Node):
+                def fn(input: Input):
+                    return self.required_parameter_1
+                return RunnableLambda(fn, name=self.__class__.__name__)
+    """
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    @classmethod
+    def build(cls, node: Node) -> Callable[[dict], dict]:
+        try:
+            built_node = cls(**node.params)
+        except ValidationError as ex:
+            raise PipelineNodeBuildError(ex)
+
+        return built_node.get_runnable(node)
+
+    @classmethod
+    def get_callable(cls, node: Node) -> Callable:
+        built_node = cls.build(node)
+
+        def fn(state: PipelineState) -> PipelineState:
+            if isinstance(built_node, BasePromptTemplate):
+                input = {"input": state["messages"][-1]}
+            else:
+                input = state["messages"][-1]
+            result = built_node.invoke(input)
+            if isinstance(result, BaseMessage):
+                output = result.content
+            elif isinstance(result, StringPromptValue):
+                output = result.text
+            else:
+                output = result
+            return PipelineState(messages=[output])
```

Comment:
Should we not append the output to the existing messages instead?

Reply:
Yes, this is very confusing, but langgraph does what you think it should here. 

The outputs are collated based on the type annotation on `PipelineState` here:
https://github.com/dimagi/open-chat-studio/blob/a73095350768bf40caf70f6afb7a01d4097c273b/apps/pipelines/nodes/base.py#L19-L21

I think because of the way langgraph can create branches that then merge together, the collating of results happens outside of this function.  

https://python.langchain.com/v0.1/docs/langgraph/#define-the-agent-state

Reply:
Ah right, thanks for the link to the docs. I saw the `add_messages` method and thought that that was used (surely), but couldn't find where, so then I assumed it wasn't. Cool cool!